File: react-native/ios/MeasureModule.swift
import Foundation
import React
import Measure

@objc(MeasureModule)
class MeasureModule: NSObject, RCTBridgeModule {
    static func moduleName() -> String! {
        return ModuleConstants.moduleName
    }
    
    static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc
    func initialize(_ clientDict: NSDictionary,
                    configDict: NSDictionary,
                    resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let clientInfo = clientDict.decoded(as: ClientInfo.self) else {
            reject(ErrorMessages.invalidArguments, "Could not decode clientDict", nil)
            return
        }
        
        guard let config = configDict.decoded(as: BaseMeasureConfig.self) else {
            reject(ErrorMessages.invalidArguments, "Could not decode configDict", nil)
            return
        }
        
        DispatchQueue.main.async {
            Measure.initialize(with: clientInfo, config: config)
        }
        resolve("Native Measure SDK initialized successfully")
    }
    
    @objc
    func start(_ resolve: @escaping RCTPromiseResolveBlock,
               rejecter reject: @escaping RCTPromiseRejectBlock) {
        Measure.start()
        resolve("Measure SDK started successfully")
    }
    
    @objc
    func stop(_ resolve: @escaping RCTPromiseResolveBlock,
              rejecter reject: @escaping RCTPromiseRejectBlock) {
        Measure.stop()
        resolve("Measure SDK stopped successfully")
    }
    
    @objc
    func trackEvent(_ data: NSDictionary,
                    type: NSString,
                    timestamp: NSNumber,
                    attributes: NSDictionary,
                    userDefinedAttrs: NSDictionary,
                    userTriggered: Bool,
                    sessionId: NSString?,
                    threadName: NSString?,
                    attachments: NSArray,
                    resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
        var mutableData = data as? [String: Any?] ?? [:]
        
        let userAttrs = userDefinedAttrs.transformAttributes()
        let msrAttachments: [MsrAttachment] = [] // TODO: map properly later
        
        Measure.internalTrackEvent(
            data: &mutableData,
            type: type as String,
            timestamp: timestamp.int64Value,
            attributes: attributes as? [String: Any?] ?? [:],
            userDefinedAttrs: userAttrs,
            userTriggered: userTriggered,
            sessionId: sessionId as String?,
            threadName: threadName as String?,
            attachments: msrAttachments
        )
        resolve("Event tracked successfully")
    }
    
    @objc
    func trackSpan(_ name: NSString,
                   traceId: NSString,
                   spanId: NSString,
                   parentId: NSString?,
                   startTime: NSNumber,
                   endTime: NSNumber,
                   duration: NSNumber,
                   status: NSNumber,
                   attributes: NSDictionary,
                   userDefinedAttrs: NSDictionary,
                   checkpoints: NSDictionary,
                   hasEnded: Bool,
                   isSampled: Bool,
                   resolver resolve: @escaping RCTPromiseResolveBlock,
                   rejecter reject: @escaping RCTPromiseRejectBlock) {
        let attrDict = attributes as? [String: Any?] ?? [:]
        let checkpointDict = checkpoints as? [String: Int64] ?? [:]
        let userAttrs = userDefinedAttrs.transformAttributes()
        
        Measure.internalTrackSpan(
            name: name as String,
            traceId: traceId as String,
            spanId: spanId as String,
            parentId: parentId as String?,
            startTime: startTime.int64Value,
            endTime: endTime.int64Value,
            duration: duration.int64Value,
            status: status.int64Value,
            attributes: attrDict,
            userDefinedAttrs: userAttrs,
            checkpoints: checkpointDict,
            hasEnded: hasEnded,
            isSampled: isSampled
        )
        resolve("Span tracked successfully")
    }

    @objc
    func setUserId(_ userId: NSString,
                   resolver resolve: @escaping RCTPromiseResolveBlock,
                   rejecter reject: @escaping RCTPromiseRejectBlock) {
        Measure.setUserId(userId as String)
        resolve("User ID set successfully")
    }

    @objc
    func clearUserId(_ resolve: @escaping RCTPromiseResolveBlock,
                     rejecter reject: @escaping RCTPromiseRejectBlock) {
        Measure.clearUserId()
        resolve("User ID cleared successfully")
    }

    @objc(trackHttpEvent:method:startTime:endTime:statusCode:error:requestHeaders:responseHeaders:requestBody:responseBody:client:resolver:rejecter:)
    func trackHttpEvent(
        url: String,
        method: String,
        startTime: NSNumber,
        endTime: NSNumber,
        statusCode: NSNumber?,
        error: String?,
        requestHeaders: [String: String]?,
        responseHeaders: [String: String]?,
        requestBody: String?,
        responseBody: String?,
        client: String,
        resolve: RCTPromiseResolveBlock,
        reject: RCTPromiseRejectBlock
    ) {
        Measure.trackHttpEvent(
            url: url,
            method: method,
            startTime: startTime.uint64Value,
            endTime: endTime.uint64Value,
            client: client,
            statusCode: statusCode?.intValue,
            error: error != nil ? NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: error!]) : nil,
            requestHeaders: requestHeaders,
            responseHeaders: responseHeaders,
            requestBody: requestBody,
            responseBody: responseBody
        )
        resolve("ok")
    }
    
    @objc
    func launchBugReport(_ takeScreenshot: Bool,
                         bugReportConfig: [String: Any],
                         attributes: [String: Any],
                         resolver: RCTPromiseResolveBlock,
                         rejecter: RCTPromiseRejectBlock) {
        Measure.launchBugReport(
            takeScreenshot: takeScreenshot,
            bugReportConfig: .default,
            attributes: attributes
        )
        resolver("Bug report launched")
    }
    
    @objc
    func setShakeListener(_ enable: Bool) {
        if enable {
            Measure.onShake {
                DispatchQueue.main.async {
                    if let bridge = RCTBridge.current(),
                       let emitter = bridge.module(for: MeasureOnShake.self) as? MeasureOnShake {
                        emitter.triggerShakeEvent()
                    }
                }
            }
        } else {
            Measure.onShake(nil)
        }
    }
    
    @objc
    func captureScreenshot(_ resolve: @escaping RCTPromiseResolveBlock,
                        rejecter reject: @escaping RCTPromiseRejectBlock) {
        
        DispatchQueue.main.async {
            guard let rootVC = RCTPresentedViewController() else {
                reject("NO_VIEW_CONTROLLER", "Unable to find current view controller", nil)
                return
            }

            Measure.captureScreenshot(for: rootVC) { attachment in
                guard let att = attachment else {
                    reject("CAPTURE_FAIL", "Failed to capture screenshot", nil)
                    return
                }

                var dict: [String: Any] = [
                    "name": att.name,
                    "type": att.type.rawValue,
                    "id": att.id,
                    "size": att.size
                ]

                if let path = att.path {
                    dict["path"] = path
                }

                if let bytes = att.bytes {
                    let base64String = bytes.base64EncodedString()
                    dict["bytes"] = base64String
                }

                resolve(dict)
            }
        }
    }

    @objc
    func captureLayoutSnapshot(_ resolve: @escaping RCTPromiseResolveBlock,
                               rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            guard let rootVC = RCTPresentedViewController() else {
                reject("NO_VIEW_CONTROLLER", "Unable to find current view controller", nil)
                return
            }

            Measure.captureLayoutSnapshot(for: rootVC) { attachment in
                guard let att = attachment else {
                    reject("CAPTURE_FAIL", "Failed to capture layout snapshot", nil)
                    return
                }

                var dict: [String: Any] = [
                    "name": att.name,
                    "type": att.type.rawValue,
                    "id": att.id,
                    "size": att.size
                ]

                if let path = att.path {
                    dict["path"] = path
                }

                if let bytes = att.bytes {
                    let base64String = bytes.base64EncodedString()
                    dict["bytes"] = base64String
                }

                resolve(dict)
            }
        }
    }

    @objc
    func getSessionId(_ resolve: @escaping RCTPromiseResolveBlock,
                     rejecter reject: @escaping RCTPromiseRejectBlock) {
        let sessionId = Measure.getSessionId()

        if let sessionId = sessionId {
            resolve(sessionId)
        } else {
            resolve(nil)
        }
    }
    
    @objc
    func trackBugReport(_ description: NSString,
                        attachments: NSArray,
                        attributes: NSDictionary,
                        resolver resolve: @escaping RCTPromiseResolveBlock,
                        rejecter reject: @escaping RCTPromiseRejectBlock) {

        let attrDict = attributes as? [String: Any] ?? [:]
        
        let msrAttachments = getMsrAttachments(attachments)
        
        Measure.trackBugReport(
            description: description as String,
            attachments: msrAttachments,
            attributes: attrDict
        )
        
        resolve("Bug report tracked")
    }
    
    func getMsrAttachments(_ attachments: NSArray) -> [MsrAttachment] {
            return attachments.compactMap { any in
                guard let dict = any as? [String: Any] else { return nil }

                guard
                    let name = dict["name"] as? String,
                    let typeStr = dict["type"] as? String,
                    let size = dict["size"] as? NSNumber,
                    let id = dict["id"] as? String
                else {
                    return nil
                }

                guard let type = AttachmentType(rawValue: typeStr) else {
                    return nil
                }

                var bytesData: Data? = nil
                var pathValue: String? = nil

                if let path = dict["path"] as? String {
                    pathValue = path
                }

                if let base64 = dict["bytes"] as? String,
                   let decoded = Data(base64Encoded: base64) {
                    bytesData = decoded
                }

                // Respect initializer preconditions
                if (bytesData == nil && pathValue == nil) ||
                   (bytesData != nil && pathValue != nil) {
                    return nil
                }

                return MsrAttachment(
                    name: name,
                    type: type,
                    size: size.int64Value,
                    id: id,
                    bytes: bytesData,
                    path: pathValue
                )
            }
        }
}

File: react-native/android/src/main/java/sh/measure/rn/MeasureModule.kt
@file:Suppress("unused")

package sh.measure.rn

import android.util.Base64
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.UiThreadUtil
import sh.measure.android.Measure
import sh.measure.android.MsrAttachment
import sh.measure.android.bugreport.MsrShakeListener
import sh.measure.android.config.ClientInfo
import sh.measure.android.config.MeasureConfig
import java.io.File

class MeasureModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = ModuleConstants.MODULE_NAME

    @ReactMethod
    fun initialize(clientMap: ReadableMap, configMap: ReadableMap, promise: Promise) {
        try {
            val context = reactContext.applicationContext

            val clientJson = MapUtils.toStringMap(clientMap)
            val configJson = MapUtils.toMap(configMap)

            val clientInfo = ClientInfo.fromJson(clientJson)
            val config = MeasureConfig.fromJson(configJson)

            UiThreadUtil.runOnUiThread {
                try {
                    Measure.init(context, measureConfig = config, clientInfo = clientInfo)
                    promise.resolve("Native Measure SDK initialized successfully")
                } catch (e: Exception) {
                    promise.reject(ErrorCode.INIT_ERROR, e)
                }
            }

        } catch (e: Exception) {
            promise.reject(ErrorCode.INIT_ERROR, "Failed to initialize Measure SDK", e)
        }
    }

    @ReactMethod
    fun start(promise: Promise) {
        Measure.start()
        promise.resolve("Measure SDK started successfully")
    }

    @ReactMethod
    fun stop(promise: Promise) {
        Measure.stop()
        promise.resolve("Measure SDK stopped successfully")
    }

    @ReactMethod
    fun trackEvent(
        data: ReadableMap,
        type: String,
        timestamp: Double,
        attributes: ReadableMap?,
        userDefinedAttrs: ReadableMap?,
        userTriggered: Boolean,
        sessionId: String?,
        threadName: String?,
        @Suppress("unused") attachments: ReadableArray?,
        promise: Promise
    ) {
        try {
            val dataMap = MapUtils.toMutableMap(data)
            val attributesMap =
                if (attributes != null) MapUtils.toMutableMap(attributes) else mutableMapOf()
            val userAttrs =
                userDefinedAttrs?.let { MapUtils.toAttributeValueMap(it) } ?: mutableMapOf()
            // TODO: parse attachments when they are supported in RN
            val attachmentsList = mutableListOf<MsrAttachment>()
            val tsLong = timestamp.toLong()

            Measure.internalTrackEvent(
                data = dataMap,
                type = type,
                timestamp = tsLong,
                attributes = attributesMap,
                userDefinedAttrs = userAttrs,
                attachments = attachmentsList,
                userTriggered = userTriggered,
                sessionId = sessionId,
                threadName = threadName
            )

            promise.resolve("Event tracked successfully")
        } catch (e: Exception) {
            promise.reject(ErrorCode.TRACK_EVENT_ERROR, "Failed to track event", e)
        }
    }

    @ReactMethod
    fun trackSpan(
        name: String,
        traceId: String,
        spanId: String,
        parentId: String?,
        startTime: Double,
        endTime: Double,
        duration: Double,
        status: Int,
        attributes: ReadableMap?,
        userDefinedAttrs: ReadableMap?,
        checkpoints: ReadableMap?,
        hasEnded: Boolean,
        isSampled: Boolean,
        promise: Promise
    ) {
        try {
            val attributesMap =
                if (attributes != null) MapUtils.toMutableMap(attributes) else mutableMapOf<String, Any?>()

            val userAttrs =
                userDefinedAttrs?.let { MapUtils.toAttributeValueMap(it) } ?: mutableMapOf<String, Any>()

            val checkpointsMap = mutableMapOf<String, Long>()

            checkpoints?.let { nonNullCheckpoints ->
                val raw = MapUtils.toMap(nonNullCheckpoints)
                raw.forEach { (key, value) ->
                    val ts = when (value) {
                        is Double -> value.toLong()
                        is Int -> value.toLong()
                        is Long -> value
                        else -> System.currentTimeMillis()
                    }
                    checkpointsMap[key] = ts
                }
            }

            Measure.internalTrackSpan(
                name = name,
                traceId = traceId,
                spanId = spanId,
                parentId = parentId,
                startTime = startTime.toLong(),
                endTime = endTime.toLong(),
                duration = duration.toLong(),
                status = status,
                attributes = attributesMap,
                userDefinedAttrs = userAttrs,
                checkpoints = checkpointsMap,
                hasEnded = hasEnded,
                isSampled = isSampled
            )

            promise.resolve("Span tracked successfully")
        } catch (e: Exception) {
            promise.reject(ErrorCode.TRACK_EVENT_ERROR, "Failed to track span", e)
        }
    }

    @ReactMethod
    fun setUserId(userId: String, promise: Promise) {
        try {
            Measure.setUserId(userId)
            promise.resolve("User ID set successfully")
        } catch (e: Exception) {
            promise.reject("SET_USER_ID_ERROR", "Failed to set userId", e)
        }
    }

    @ReactMethod
    fun clearUserId(promise: Promise) {
        try {
            Measure.clearUserId()
            promise.resolve("User ID cleared successfully")
        } catch (e: Exception) {
            promise.reject("CLEAR_USER_ID_ERROR", "Failed to clear userId", e)
        }
    }

    @ReactMethod
    fun trackHttpEvent(
        url: String,
        method: String,
        startTime: Double,
        endTime: Double,
        statusCode: Int?,
        error: String?,
        requestHeaders: ReadableMap?,
        responseHeaders: ReadableMap?,
        requestBody: String?,
        responseBody: String?,
        client: String,
        promise: Promise
    ) {
        val errObj = error?.let { Exception(it) }

        Measure.trackHttpEvent(
            url,
            method,
            startTime.toLong(),
            endTime.toLong(),
            statusCode,
            errObj,
            requestHeaders?.toHashMap()?.mapValues { it.value.toString() }?.toMutableMap(),
            responseHeaders?.toHashMap()?.mapValues { it.value.toString() }?.toMutableMap(),
            requestBody,
            responseBody,
            client
        )

        promise.resolve("ok")
    }

    @ReactMethod
    fun launchBugReport(
        takeScreenshot: Boolean,
        bugReportConfig: ReadableMap?,
        attributes: ReadableMap?,
        promise: Promise
    ) {
        try {
            val userAttrs =
                attributes?.let { MapUtils.toAttributeValueMap(it) } ?: mutableMapOf()
            Measure.launchBugReportActivity(takeScreenshot, userAttrs)

            promise.resolve("Bug report launched successfully")
        } catch (e: Exception) {
            promise.reject("LAUNCH_BUG_REPORT_FAILED", e)
        }
    }

    @ReactMethod
    fun setShakeListener(enable: Boolean) {
        val shakeEmitter = reactApplicationContext
            .getNativeModule(MeasureOnShakeModule::class.java)

        if (enable) {
            Measure.setShakeListener(object : MsrShakeListener {
                override fun onShake() {
                    shakeEmitter?.triggerShakeEvent()
                }
            })
        } else {
            Measure.setShakeListener(null)
        }
    }

   @ReactMethod
    fun captureScreenshot(promise: Promise) {
        val activity = currentActivity
        if (activity == null) {
            promise.reject("NO_ACTIVITY", "No current activity available")
            return
        }

        UiThreadUtil.runOnUiThread {
            try {
                Measure.captureScreenshot(
                    activity,
                    onComplete = { attachment ->

                        try {
                            val map = Arguments.createMap().apply {
                                putString("id", attachment.name)
                                putString("name", attachment.name)
                                putString("type", attachment.type)

                                attachment.path?.let { filePath ->
                                    putString("path", filePath)

                                    val file = File(filePath)
                                    if (file.exists()) {
                                        putInt("size", file.length().toInt())
                                    }
                                }

                                attachment.bytes?.let { bytes ->
                                    val base64 = Base64.encodeToString(bytes, Base64.NO_WRAP)
                                    putString("bytes", base64)
                                    putInt("size", bytes.size)
                                }
                            }

                            promise.resolve(map)
                        } catch (e: Exception) {
                            promise.reject("MAP_ERROR", "Failed building result map", e)
                        }
                    },
                    onError = {
                        promise.reject("CAPTURE_FAIL", "Failed to capture screenshot")
                    }
                )
            } catch (e: Exception) {
                promise.reject("SCREENSHOT_CAPTURE_FAILED", e)
            }
        }
    }

    @ReactMethod
    fun captureLayoutSnapshot(promise: Promise) {
        val activity = currentActivity
        if (activity == null) {
            promise.reject("NO_ACTIVITY", "No current activity available")
            return
        }

        UiThreadUtil.runOnUiThread {
            try {
                Measure.captureLayoutSnapshot(
                    activity,
                    onComplete = { attachment ->
                        try {
                            val map = Arguments.createMap().apply {
                                putString("id", attachment.name)
                                putString("name", attachment.name)
                                putString("type", attachment.type)

                                attachment.path?.let { filePath ->
                                    putString("path", filePath)

                                    val file = File(filePath)
                                    if (file.exists()) {
                                        putInt("size", file.length().toInt())
                                    }
                                }

                                attachment.bytes?.let { bytes ->
                                    val base64 = Base64.encodeToString(bytes, Base64.NO_WRAP)
                                    putString("bytes", base64)
                                    putInt("size", bytes.size)
                                }
                            }

                            promise.resolve(map)

                        } catch (e: Exception) {
                            promise.reject("MAP_ERROR", "Failed building snapshot map", e)
                        }
                    },
                    onError = {
                        promise.reject("LAYOUT_SNAPSHOT_FAIL", "Failed to capture layout snapshot")
                    }
                )
            } catch (e: Exception) {
                promise.reject("LAYOUT_SNAPSHOT_EXCEPTION", e)
            }
        }
    }

    @ReactMethod
    fun trackBugReport(
        description: String,
        attachments: ReadableArray?,
        attributes: ReadableMap?,
        promise: Promise
    ) {
        try {
            val attrs =
                attributes?.let { MapUtils.toAttributeValueMap(it) } ?: mutableMapOf()

            val msrAttachments = getMsrAttachments(attachments)

            Measure.trackBugReport(
                description,
                msrAttachments,
                attrs
            )

            promise.resolve("Bug report tracked successfully")
        } catch (e: Exception) {
            promise.reject("TRACK_BUG_REPORT_FAILED", e)
        }
    }

    private fun getMsrAttachments(attachments: ReadableArray?): List<MsrAttachment> {
        if (attachments == null || attachments.size() == 0) return emptyList()

        val list = mutableListOf<MsrAttachment>()

        for (i in 0 until attachments.size()) {
            val map: ReadableMap? = attachments.getMap(i)
            if (map == null) continue  // <-- FIXED NULL CHECK

            val name = map.getString("name") ?: continue
            val type = map.getString("type") ?: continue
            val path = map.getString("path")
            val base64 = map.getString("bytes")

            // If 'path' exists, use it.
            // If 'bytes' exists, decode & write to temp file.
            val finalPath = when {
                path != null -> path

                base64 != null -> {
                    try {
                        val bytes = Base64.decode(base64, Base64.DEFAULT)
                        val tempFile = File(reactContext.cacheDir, name)
                        tempFile.writeBytes(bytes)
                        tempFile.absolutePath
                    } catch (_: Exception) {
                        null
                    }
                }

                else -> null
            }

            if (finalPath != null) {
                list.add(
                    MsrAttachment(
                        name = name,
                        path = finalPath,
                        type = type,
                    )
                )
            }
        }

        return list
    }
}